---
title: "Week1 Homework"
output: html_notebook
---

---
**Qn1.Using no other R functions besides sqrt (e.g. to calculate sum use a loop summing all elements from first to last) implement function my.sd calculating standard deviation for a numeric vector of arbitrary length and demonstrate that it produces exactly the same results as R function sd.**

Steps:
1. create function
2. check for NAs
3. find mean
4. find deviation for each data point (datapoint - mean)
5. square each deviation
6. find mean of pooled deviations (when calculating use len -1) = variance
7. sqrt the variance
8. test to compare to the std function

```{r}
my.sd <- function(x, rm.na = T) {
  x.clean <- c() #initiate clean vector - without NAs
  x.sum <- 0
  x.len <- 0
  for (xx in x) {
    if (is.na(xx)) { #check for NAs
      if (rm.na) {
        next
      } else {
        return (NA)
      }
    } 
    x.clean <- c(x.clean, xx) 
    x.sum <- x.sum + xx #sum of all non na values
    x.len <- x.len +1 # find length of vector minus NAs
  }
  x.mean <- x.sum / x.len
  x.devsq <- (x.clean -x.mean)^2 #subtract mean from each value in vector and square
  
  x.devsq.sum <- 0 
  for (xd in x.devsq) x.devsq.sum <- x.devsq.sum + xd
  
  x.var = x.devsq.sum / (x.len -1) #-1 for sample sd
  x.std <- sqrt(x.var)
  
  #list(x.clean, x.len, x.sum, x.mean, x.devsq, x.devsq.sum, x.var, x.std)
  x.std
}

v <- c(1,2,4,9, NA)
my.sd(v)
sd(v, na.rm=T)
```
**2.Execute and explain the results of the following three pieces of R code. For each of them – start R afresh, copy and paste code in R command prompt. Explain in detail what the code does, what is printed (or not printed) and why.**
**a.**
```{r}
x<-sin(1);print(x);y<-Sin(2);print(y)
x<-sin(1)
print(x)
y<-Sin(2)
print(y)
```
When executed:
```{r}
x<-sin(1);print(x);y<-Sin(2);print(y)
x<-sin(1)
print(x)
y<-Sin(2)
print(y)
```

This code takes the sin of 1, assigns it to the variable x, prints it to the screen then calls an error. This error occurs in the assignment of Sin(2) to variable y. the correct function is sin() not Sin() and this typo has therefore stalled the rest of the code from running. Looking at the remainder of this code, the 1st line completes the same task as lines 2-5 put together, it has just been condensed into 1 line as `;` can be used to separate commands.

**b.**
```{r}
m <- matrix(c(1:5,NA,7,7,NA),
nrow=3,ncol=3,byrow=T)
print(m)
for ( i in 1:dim(m)[[1]] ) {
 mm <- sort(m[i,],na.last=c(NA,NA,T)[[i]])
 for ( j in 1:(length(mm)-1) ) {
 if ( mm[[j]]==mm[[j+1]] ) {
 cat(i,j,mm[[j]],fill=T)
 }
 }
}
```

When executed:
```{r}
m <- matrix(c(1:5,NA,7,7,NA),
nrow=3,ncol=3,byrow=T)
print(m)
for ( i in 1:dim(m)[[1]] ) {
 mm <- sort(m[i,],na.last=c(NA,NA,T)[[i]])
 for ( j in 1:(length(mm)-1) ) {
 if ( mm[[j]]==mm[[j+1]] ) {
 cat(i,j,mm[[j]],fill=T)
 }
 }
}
```
Tough one to get my head around. What I believe the code is supposed to do is creating a 3x3 matrix from the vector of values and inputting it into the matrix by rows. It then goes into a outer for loop that iterates over each row (as diam() results in row x column and we select the 1st of the two with [[1]]). For each row it attempts to sort the values in ascending order and includes the na.last argument which dictates how to handle NA values in the vector. Per the help page, "If TRUE, missing values in the data are put last; if FALSE, they are put first; if NA, they are removed". A vector of (NA,NA,TRUE)[[i]] is passed into the na.last argument indicating that for the first and second rows na.last = NA (so NAs are removed from the row when sorted), and the third row na.last = True (NA is put last in the vector after sorting). Moving to the inner loop, which iterates over the values in each [i] row except for the last value, it checks to see if the value of that position is equal to the value of the next position in the same row (hence the -1 or the last value would be trying to compare to an index out of range). If the values are equal it prints the row number, column index position, and the duplicate value.

The issue we are getting running this code as depicted, is that in the third row na.last=T, so the NA remains in the vector to be iterated upon in the inner loop, and therefore is passed into the `mm[[j]]==mm[j+i]` equation. Because the value is missing, one cannot determine whether the value adjacent to it (7) is equal or not because the NA value is missing. To fix this, we can change the sort function to look like this `sort(m[i,],na.last=NA)`

**c.**
```{r}
1:3 * 5:7
1:3 %*% 5:7
1:3 %*% t(5:7)
```
The first line performs element-wise multiplication with `*`, so we end up with a vector.
The second line uses matrix multiplication, and the arguments are both vectors of the same length, therefore it returns the inner product (dot product), and does so as a matrix. This means it performs element-wise multiplication, then adds the products together. 
The last line is another example of matrix multiplication. I remember from the notes that matrices are stored internally as column-major order vectors. Therefore when the second vector is transposed, it promotes it to a row vector. Now we result in a matrix where the first row is the row vector multiplied by the first value in the column vector `(5:7 *3)`, the second row is the row vector multiplied by the second value in the column vector `(5:7 * 2)`, and the last row is the row vector multiplied by the third value in the column vector `(5:7 *3)`.

**3. Practice dataset import/export a. Load into R dataset aaMap (available in Biobase), determine and describe data type of each column. b. Store, remove and recover it using save and load functions (use the help command ‘?’ and read documentation), demonstrate its removal and recovery, using ls function. c. Save it in a file using write.table function d. Open the file using text editor, add a column describing amino acid side chain charge at pH=7 (you can use wikipedia in order to get neutral/positive/negative charge annotations, or you can even come up with some random numbers, this exercise is about modifying the data and loading them back, not about amino acids), save file, import it back into R, demonstrate that column was added to the data frame.**

```{r}
library(Biobase)
data("aaMap")
colnames(aaMap)
dim(aaMap)
aaMap[1:5,]

class(aaMap)
unlist(lapply(aaMap[,1:6],class))
```
The aaMap dataset from Biobase is a dataframe with 20 rows and 6 columns. Each row is an amino acid and the column names are listed above. Looking at the help page I got some more information and the descriptors of the columns are as follows:
name: amino acid name
let.1: one-letter code
let.3: three-letter code
scProp: side chain property at pH 7 (polar/nonpolar)
hyPhilic: side chain is hydrophilic at pH 7
acidic: side chain is acidic at pH 7

We can see that the first 4 columns are factors - this makes sense as they are categorical variables, and the data they hold is finite and fixed. At first I thought that the name column should be a character class, but considering there are also a fixed and finite number of amino acids, that factor is also appropriate. Columns 5 and 6 have logical classes as their columns are asking specific questions (is the side chain hydrophilic at pH 7? Is the side chain acidic at pH 7?). 

```{r}
ls()
save(aaMap,file = "my.aaMap.rda")
```
```{r}
rm(aaMap)
ls()

```
```{r}
load("my.aaMap.rda")
ls()
```
```{r}
write.table(aaMap, "aatable.txt", quote=F, sep="\t", col.names=T, row.names=F)
```

```{r}
updated.aaMap <- read.table("aatable2.txt", sep="\t", header=T)
updated.aaMap[1:5,]
sapply(updated.aaMap,class)
```
Change classes of cols 1:4,7
```{r}
updated.aaMap$name <- as.factor(updated.aaMap$name)
updated.aaMap$let.1 <- as.factor(updated.aaMap$let.1)
updated.aaMap$let.3 <- as.factor(updated.aaMap$let.3)
updated.aaMap$scProp <- as.factor(updated.aaMap$scProp)
updated.aaMap$charge <- as.factor(updated.aaMap$charge)
sapply(updated.aaMap,class)
```
**4. Calculate and visualize some basic properties of gene expression levels in ALL dataset. Remember that function exprs returns a matrix of gene expression levels (~13K genes) for each patient (>100 patients) in the sample.**

