---
title: Week 1 Homework
author: "Author: Laura Wooten"
date: "Last update: `r format(Sys.time(), '%d %B, %Y')`" 
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 4
    fig_caption: yes
    code_folding: hide
    number_sections: true
  pdf_document:
    toc: true
vignette: >
    %\VignetteIndexEntry{Text}
    %\usepackage[utf8]{inputenc}
    %\VignetteEngine{knitr::rmarkdown}
fontsize: 15pt
editor_options: 
  chunk_output_type: console
---

<style>
pre code, pre, code {
  white-space: pre !important;
  overflow-x: auto !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
body {
text-align: justify}
</style>

---
title: "Week1 Homework"
author: 'Author: Laura Wooten'
date: 'Date: 08 April 2025'
output:
  html_document:
    df_print: paged
fontsize: 15pt
editor_options:
  chunk_output_type: inline
---

---
**1. Using no other R functions besides sqrt (e.g. to calculate sum use a loop summing all elements from first to last) implement function my.sd calculating standard deviation for a numeric vector of arbitrary length and demonstrate that it produces exactly the same results as R function sd.**

Steps:  
- create function  
- check for NAs  
- find mean  
- find deviation for each data point (datapoint - mean)  
- square each deviation  
- find mean of pooled deviations (when calculating use len -1) = variance  
- sqrt the variance  
- test to compare to the std function  

```{r}
my.sd <- function(x, rm.na = T) {
  x.clean <- c() #initiate clean vector - without NAs
  x.sum <- 0
  x.len <- 0
  for (xx in x) {
    if (is.na(xx)) { #check for NAs
      if (rm.na) {
        next
      } else {
        return (NA)
      }
    } 
    x.clean <- c(x.clean, xx) 
    x.sum <- x.sum + xx #sum of all non na values
    x.len <- x.len +1 # find length of vector minus NAs
  }
  x.mean <- x.sum / x.len
  x.devsq <- (x.clean -x.mean)^2 #subtract mean from each value in vector and square
  
  x.devsq.sum <- 0 
  for (xd in x.devsq) x.devsq.sum <- x.devsq.sum + xd
  
  x.var = x.devsq.sum / (x.len -1) #-1 for sample sd
  x.std <- sqrt(x.var)
  
  #list(x.clean, x.len, x.sum, x.mean, x.devsq, x.devsq.sum, x.var, x.std)
  x.std
}

v <- c(1,2,4,9, NA)
my.sd(v)
sd(v, na.rm=T)
```
**2.Execute and explain the results of the following three pieces of R code. For each of them – start R afresh, copy and paste code in R command prompt. Explain in detail what the code does, what is printed (or not printed) and why.**  
**a.**
```{r, eval=FALSE}
x<-sin(1);print(x);y<-Sin(2);print(y)
x<-sin(1)
print(x)
y<-Sin(2)
print(y)
```
When executed:
```{r, error=TRUE}
x<-sin(1);print(x);y<-Sin(2);print(y)
x<-sin(1)
print(x)
y<-Sin(2)
print(y)
```

This code takes the sin of 1, assigns it to the variable x, prints it to the screen then calls an error. This error occurs in the assignment of Sin(2) to variable y. the correct function is sin() not Sin() and this typo has therefore stalled the rest of the code from running. Looking at the remainder of this code, the 1st line completes the same task as lines 2-5 put together, it has just been condensed into 1 line as `;` can be used to separate commands. 

**b.**
```{r, eval=FALSE}
m <- matrix(c(1:5,NA,7,7,NA),
nrow=3,ncol=3,byrow=T)
print(m)
for ( i in 1:dim(m)[[1]] ) {
 mm <- sort(m[i,],na.last=c(NA,NA,T)[[i]])
 for ( j in 1:(length(mm)-1) ) {
 if ( mm[[j]]==mm[[j+1]] ) {
 cat(i,j,mm[[j]],fill=T)
 }
 }
}
```

When executed:
```{r, error=TRUE}
m <- matrix(c(1:5,NA,7,7,NA),
nrow=3,ncol=3,byrow=T)
print(m)
for ( i in 1:dim(m)[[1]] ) {
 mm <- sort(m[i,],na.last=c(NA,NA,T)[[i]])
 for ( j in 1:(length(mm)-1) ) {
 if ( mm[[j]]==mm[[j+1]] ) {
 cat(i,j,mm[[j]],fill=T)
 }
 }
}
```
Tough one to get my head around. What I believe the code is supposed to do is creating a 3x3 matrix from the vector of values and inputting it into the matrix by rows. It then goes into a outer for loop that iterates over each row (as diam() results in row x column and we select the 1st of the two with [[1]]). For each row it attempts to sort the values in ascending order and includes the na.last argument which dictates how to handle NA values in the vector. Per the help page, "If TRUE, missing values in the data are put last; if FALSE, they are put first; if NA, they are removed". A vector of (NA,NA,TRUE)[[i]] is passed into the na.last argument indicating that for the first and second rows na.last = NA (so NAs are removed from the row when sorted), and the third row na.last = True (NA is put last in the vector after sorting). Moving to the inner loop, which iterates over the values in each [i] row except for the last value, it checks to see if the value of that position is equal to the value of the next position in the same row (hence the -1 or the last value would be trying to compare to an index out of range). If the values are equal it prints the row number, column index position, and the duplicate value.

The issue we are getting running this code as depicted, is that in the third row na.last=T, so the NA remains in the vector to be iterated upon in the inner loop, and therefore is passed into the `mm[[j]]==mm[j+i]` equation. Because the value is missing, one cannot determine whether the value adjacent to it (7) is equal or not because the NA value is missing. To fix this, we can change the sort function to look like this `sort(m[i,],na.last=NA)`

**c.**
```{r}
1:3 * 5:7
1:3 %*% 5:7
1:3 %*% t(5:7)
```
The first line performs element-wise multiplication with `*`, so we end up with a vector.
The second line uses matrix multiplication, and the arguments are both vectors of the same length, therefore it returns the inner product (dot product), and does so as a matrix. This means it performs element-wise multiplication, then adds the products together. 
The last line is another example of matrix multiplication. I remember from the notes that matrices are stored internally as column-major order vectors. Therefore when the second vector is transposed, it promotes it to a row vector. Now we result in a matrix where the first row is the row vector multiplied by the first value in the column vector `(5:7 *3)`, the second row is the row vector multiplied by the second value in the column vector `(5:7 * 2)`, and the last row is the row vector multiplied by the third value in the column vector `(5:7 *3)`.
```{r include=FALSE}
rm(list = ls())
```

**3. Practice dataset import/export a. Load into R dataset aaMap (available in Biobase), determine and describe data type of each column. b. Store, remove and recover it using save and load functions (use the help command ‘?’ and read documentation), demonstrate its removal and recovery, using ls function. c. Save it in a file using write.table function d. Open the file using text editor, add a column describing amino acid side chain charge at pH=7 (you can use wikipedia in order to get neutral/positive/negative charge annotations, or you can even come up with some random numbers, this exercise is about modifying the data and loading them back, not about amino acids), save file, import it back into R, demonstrate that column was added to the data frame.**

```{r}
library(Biobase)
data("aaMap")
colnames(aaMap)
dim(aaMap)
aaMap[1:5,]

class(aaMap)
unlist(lapply(aaMap[,1:6],class))
```
The aaMap dataset from Biobase is a dataframe with 20 rows and 6 columns. Each row is an amino acid and the column names are listed above. Looking at the help page I got some more information and the descriptors of the columns are as follows:
name: amino acid name
let.1: one-letter code
let.3: three-letter code
scProp: side chain property at pH 7 (polar/nonpolar)
hyPhilic: side chain is hydrophilic at pH 7
acidic: side chain is acidic at pH 7

We can see that the first 4 columns are factors - this makes sense as they are categorical variables, and the data they hold is finite and fixed. At first I thought that the name column should be a character class, but considering there are also a fixed and finite number of amino acids, that factor is also appropriate. Columns 5 and 6 have logical classes as their columns are asking specific questions (is the side chain hydrophilic at pH 7? Is the side chain acidic at pH 7?). 

```{r}
ls()
save(aaMap,file = "my.aaMap.rda")
```
```{r}
rm(aaMap)
ls()

```
```{r}
load("my.aaMap.rda")
ls()
```
```{r}
write.table(aaMap, "aatable.txt", quote=F, sep="\t", col.names=T, row.names=F)
```

```{r}
updated.aaMap <- read.table("aatable2.txt", sep="\t", header=T)
updated.aaMap[1:5,]
sapply(updated.aaMap,class)
```
Change classes of cols 1:4,7
```{r}
updated.aaMap$name <- as.factor(updated.aaMap$name)
updated.aaMap$let.1 <- as.factor(updated.aaMap$let.1)
updated.aaMap$let.3 <- as.factor(updated.aaMap$let.3)
updated.aaMap$scProp <- as.factor(updated.aaMap$scProp)
updated.aaMap$charge <- as.factor(updated.aaMap$charge)
sapply(updated.aaMap,class)
```
**4. Calculate and visualize some basic properties of gene expression levels in ALL dataset. Remember that function exprs returns a matrix of gene expression levels (~13K genes) for each patient (>100 patients) in the sample.**

**a. Calculate average gene expression level for each patient. Plot them using the following visualizations: i. Histogram ii. Boxplot iii. Stripchart iv. Stem-and-leaf v. Dotchart vi. Sorted in ascending order**

```{r}
library(ALL)
data(ALL)
clinical <- pData(ALL)
geneEx <- exprs(ALL)

```

```{r}
dim(geneEx)
geneEx[1:5,1:5]
```

The exprs() help page says "Matrix with rows representing features and columns samples", therefore the patients are our columns, and genes are rows. Given this we can find the mean of each column (total 128) using vectors (as vectors are column-major order in matrices).

```{r}
patient.means <- sapply(1:128, function(i) mean(geneEx[,i]))
patient.means[1:20] #preview results
```
```{r}
hist(patient.means, main=paste("Histogram of Average Patient Gene Expression Levels"), xlab ="Average Expression Levels")
```

```{r}
boxplot(patient.means, ylab="Patient Average Expression Levels")
```

```{r}
stripchart(patient.means,method = "jitter", xlab="Patient Average Expression Levels")
```

```{r}
stem(patient.means)
dotchart(patient.means, xlab="Patient Average Expression Levels")
```

```{r}
plot(sort(patient.means),type="p", main="Dot Plot of Average Patient Gene Expression Levels", ylab = "Average Expression Levels", xlab="Patients")
```
**b. Calculate average gene expression levels for each gene. Plot them using: i. Histogram ii. Boxplot iii. sorted in the ascending order**

```{r}

# gene.means <- sapply(1:12625, function(i) mean(t(geneEx)[,i])) 
gene.means <- rowMeans(geneEx)
#preview
gene.means[1:10]
hist(gene.means, main=paste("Histogram of Average Gene Expression Levels, per Gene"), xlab ="Average Expression Levels")
```
Originally thought to transpose the matrix then use sapply() to the columns, but given this results in 12625 columns it was too inefficient. Found the rowMeans() function through a quick google search.
```{r}
boxplot(gene.means, ylab="Average Expression Levels")
```

```{r}
plot(sort(gene.means),type="p", main="Dot Plot of Average Gene Expression Levels, per Gene", ylab = "Average Expression Levels", xlab="Gene index")
```
**c. For average gene expression per patient calculate the following measures of center and spread – explain results obtained: i. Mean ii. Median iii. Standard deviation (sd) iv. Median absolute deviation (mad) v. Interquartile range vi. Five number summary (read the docs for, and use fivenum() function) vii. Five number summary using quantile function**
```{r}
mean.value <- mean(patient.means)
med.value <- median(patient.means)
sd.value <- sd(patient.means)
mad.value <- mad(patient.means)
iqr.value <- IQR(patient.means)
fn.value <- fivenum(patient.means)
quant.value <- quantile(patient.means)

cat("Mean:", mean.value, "\n",
    "Median:", med.value, "\n",
    "Standard Deviation:", sd.value, "\n",
    "Median Absolute Deviation:", mad.value, "\n",
    "Interquartile Range:", iqr.value, "\n",
    "Five Number Summary (fivenum):", fn.value, "\n",
    "Five Number Summary (quantile):", quant.value, "\n")

```
The average gene expression level across all patients is 5.624988. The middle value of all gene expression values in 5.626497 - this means that half of the patients have average expression levels below this number. The standard deviation measures the amount of variation from the mean. In this case the SD is 0.0226, this is a very small number indicating that gene expression levels across the sample are all very close to the mean. The MAD is another measurement to assess variation across the data, however it is less sensitive to outliers as it uses the median instead of the mean to calculate variation. The IQR measures the spread of the central 50% of data, in this case between (5.62-0.028) - (5.62 + 0.028). The fivenum() and quantile() functions both give an overview of the distribution of the data, but they use different methods to calculate the quantiles - fivenum() divides the data into four equal parts, while quantile() uses interpolation to calculate the quantiles. 

**d. For average gene expression per gene calculate the same measures of center and spread as above – explain results obtained.**
```{r}
mean.g <- mean(gene.means)
med.g <- median(gene.means)
sd.g <- sd(gene.means)
mad.g <- mad(gene.means)
iqr.g <- IQR(gene.means)
fn.g <- fivenum(gene.means)
quant.g <- quantile(gene.means)

cat("Mean:", mean.g, "\n",
    "Median:", med.g, "\n",
    "Standard Deviation:", sd.g, "\n",
    "Median Absolute Deviation:", mad.g, "\n",
    "Interquartile Range:", iqr.g, "\n",
    "Five Number Summary (fivenum):", fn.g, "\n",
    "Five Number Summary (quantile):", quant.g, "\n")
```
Like with the average patient expression levels, the mean of average gene expression level is 5.624988, however the middle value when all gene average expression levels are ordered is slightly lower at 5.49578. The standard deviation and MAD are also larger in this analysis, indicating that there is greater variation in expression levels per gene then there is per patient. The middle 50% of average gene expression levels fall between 4.17- 6.80, and the lowest and highest average expression levels of genes are 2.59 and 13.57, respectively. 

**e. Find genes with the highest and the lowest expression levels across patients – capture their names and characteristic expression levels. For this portion of the question, you could look either for genes with lowest/highest average expression or for the ones with lowest/highest single expression measurement across all patients. However, the average expression level is more biologically meaningful, as a single observation is more likely to be an error/outlier. Next, Calculate for each of them the above measures of center and spread across patients. Plot distributions of their expression levels across patients using all visualization techniques practiced above.**

```{r}
lowest.gene.level <- quant.g[1]; lowest.gene.level
highest.gene.level <- quant.g[5]; highest.gene.level

#find which gene correlates to these averages from gene.means vector

lowest.gene <- names(gene.means[lowest.gene.level]); lowest.gene
highest.gene <- names(gene.means[highest.gene.level]); highest.gene

```
```{r}
#input this into geneEx matrix, find patient data for each gene

lg.patients <- geneEx[lowest.gene,] #row lowest gene, all columns
hg.patients <- geneEx[highest.gene,] #select row highest gene, all columns
```

To create summary statistics and visualizations of each gene, I'll attempt to create a function that I can use each vector as an argument.
```{r}
stats.and.vis <- function(vec, gene.name){
  #create inner function to calc summary stats
  calc.stats <- function(vec) {
    stats <- list(
      mean = mean(vec),
      median = median(vec),
      sd = sd(vec),
      mad = mad(vec),
      iqr = IQR(vec),
      five.num.summary = fivenum(vec),
      quantile.summary = quantile(vec)
    )
    return(stats)
  }
  #call calc.stats and print 
  stats <- calc.stats(vec)
  cat("Statistics for ", gene.name, ": \n")
  print(stats)
    
  #plot distributions
  hist(vec, main=paste("Histogram of ", gene.name, " Expression Levels across Patients"), xlab="Expression level")
  boxplot(vec, main=paste("Boxplot of ", gene.name, " Expression Levels across Patients"))
  stripchart(vec, main=paste("Stripchart of ", gene.name, " Expression Levels across Patients"),method = "jitter")
  stem(vec)
  dotchart(vec, main=paste("Dot chart of ", gene.name, " Expression Levels across Patients"))
  plot(sort(vec),type="p", main=paste("Ordered Dot Plot of ", gene.name, " Expression Levels across Patients"), ylab = "Expression Level", xlab="Patients")
  
}
```

Call function for the lowest and highest average expression genes.  
Lowest (1001_at):
```{r}
stats.and.vis(lg.patients, lowest.gene)
```

Highest (1011_s_at):
```{r}
stats.and.vis(hg.patients, highest.gene)
```
**f. Qualitatively compare distributions of mean gene expression levels per gene and per patient. Describe results of such comparison in plain English.**

Between the statistics and visualizations of the two genes, the obvious difference is that the mean and median of 1011_a_at are higher than those of 1001_at. This is completely expected as I pulled these genes from the initial matrix as the highest and lowest averagely expressed genes across patients from the ALL dataset. The variance and distribution of 1011_a_at are greater and more spread out, with a higher standard deviation and interquartile range, which indicates there are more diverse expression levels among patients compared to 1001_at, which is tightly packed. This means that the expression of 1001_at is fairly consistent among patients despite different clinical traits of the disease.

Something of note is that the high-range outliers of 1011_a_at seem to be grouped together (i.e., there are a few at the same approximate value) compared to 1001_at. This could indicate that there are subsets of patients that may share characteristics or clinical traits of the disease. Looking at the box plot, the outliers are closer to the whiskers in 1011_a_at, and the ordered dot plot has less of a gap in the curve. This may indicate that these high expression levels are not extreme anomalies but instead part of a broader pattern within the patient population. This is also demonstrated in the histograms of both; 1011_a_at shows a normal (bell curve) distribution, while 1001_at is left-skewed.

If this analysis were to go one step further and integrate with the clinical data present in the ALL dataset, it could reveal correlations between gene expression levels and specific clinical outcomes or patient characteristics. This could help identify potential biomarkers for prognosis or targets for personalized treatment strategies in ALL patients.


