---
title: Week 6 Homework
author: "Author: Laura Wooten"
date: "Last update: `r format(Sys.time(), '%d %B, %Y')`" 
output: 
  html_notebook:
    code_folding: 
    number_sections: true
---

```{r}
library(GEOquery)
geodata <-  getGEO("GSE27272", GSEMatrix = TRUE)
eset <- geodata[[1]]
edata <- exprs(eset)
pheno <- pData(eset)
```
```{r}
#Save full data set
save(edata,file = "cotinine_edata.rda")
save(pheno, file = "cotinine_pheno.rda")
```


Data Processing - "Bead summary data were imported into the R statistical environment (www.r-project.org) and normalized by the quantile method in the Lumi package. Only probes that reached detection P-value < 0.01 in all samples were used for further analyses. Differentially expressed genes were analysed in the Limma package. A linear model was fitted for each gene given a series of arrays using the lmFit function. Using the Benjamini and Hochberg method, P-values were corrected for multiple testing. The M, PL, and D groups were normalized separately."

Check normalization of array data - Plot average expression to check for normalized data - group samples by levels of cotinine concentration, look at spread of averaged values.

```{r}
#check for normalization of data
#"cc_conc" is a vector of cord cotinine concentrations (independent variable)
cc_conc <- as.numeric(gsub(".*: ","",pheno$characteristics_ch1.11))
# And "edata" is a matrix where rows = genes, columns = samples (expression levels)

# Define cotinine exposure groups
quartiles <- quantile(cc_conc, probs = c(0.25, 0.50, 0.75)) # Compute quartiles
exposure_group <- cut(cc_conc, 
                      breaks = c(-Inf, quartiles[1], quartiles[3], Inf), 
                      labels = c("Low", "Moderate", "High"))

# Compute average gene expression per sample
avg_expression <- colMeans(edata)

# Create a dataframe
df <- data.frame(Sample = colnames(edata), 
                 AvgExpression = avg_expression, 
                 CotinineGroup = exposure_group)

# Box plot stratified by cotinine levels
boxplot(AvgExpression ~ CotinineGroup, data = df, 
        main = "Average Gene Expression Stratified by Cotinine Levels",
        xlab = "Cotinine Exposure Group", 
        ylab = "Average Gene Expression",
        col = c("lightblue", "lightgreen", "salmon"))

boxplot(edata, main = "Gene Expression Levels Across Samples", las = 2)  # 'las = 2' rotates labels

```


To assess that normalization of the data was successful, I averaged the gene expression levels across all genes for each sample/patient. Each sample was then grouped by the independent variable as either low, medium or high cotinine levels (lowest 25%, IQR, highest 25%). The results of the average gene expressions per group can be seen on the box plot above. Here we see that the medians and overall distributions look the same between all groups. This is what we'd expect when our data has been normalized because quantile normalization (as used in the protocol for this data set) adjusts all samples so that the overall distribution of samples is equalized. This removes biases die to technical variability, as the overall median we are seeing for each group is a summary of 20k+ genes, we are investigating how the dependent variable affects individual gene expression, and don't expect a difference in the overall average level of all genes in the sample.

```{r include=FALSE}
rm(list=setdiff(ls(), c("edata","pheno","cc_conc")))
```


> 1. 30 points With the downloaded data set, fit a linear model using the continuous annotation as the dependent variable and each gene expression as the dependent variable. Then calculate p-values for each linear model fit using an anova. This should generate a p-value for each gene. Next, calculate a corrected p-value using the Bonferoni correction and one using a FDR calculation. Annotate all three of these p-value calculations onto the data table that was downloaded. Next, generate scatter plots with genes ordered by the significance level for each independent p-value measurement, and the -log10(p-value) on the y-axis. Describe all these findings in addition to any changes in gene order by significance that are observed by the different p-value adjustments.


So originally I planned to use cotinine concentration as the independent variable and see how it affected gene expression, but as this question specifically states to use gene expression as the dependent variable, I'll flip the way I view the data to account for this.
**Define Variables:**  
Dependent continuous variable = "characteristics_ch1.11" cord blood cotinine (ng/ml).
Independent variable = expression level.

1. Partition data into smoking category.  
2. Use apply() for lm + anova over each row. 
3. extract p-value, calc bonferoni + FDR, add both to df   
3. Scatter plots   
4. Describe
```{r}
#pheno (phenotype information for samples) shows there are 3 tissue samples per patient, we only want to look at samples from cord blood > limit it so we only have 1 sample per patient. Also want to partion for smoker/nonsmoker

#limit data to cord blood samples only
pheno.cord <- pheno[pheno$source_name_ch1=="cord blood, smoker" | pheno$source_name_ch1=="cord blood, non-smoker",]
edata.cord <- edata[,row.names(pheno.cord)]

#Find accessions for sample: cord blood, outcome: smoker
#cord_smoker_acc <- row.names(pheno[pheno$source_name_ch1=="cord blood, smoker",]["geo_accession"])
#Find accessions for sample: cord blood, outcome: non-smoker
#cord_nonsm_acc <- row.names(pheno[pheno$source_name_ch1=="cord blood, non-smoker",]["geo_accession"])

#partition edata by cord_smoker_acc and cord_nonsm_acc
#edata.smoker <- edata[,cord_smoker_acc]
#edata.nonsm <- edata[,cord_nonsm_acc]

#get cotinine cord blood concentration from pheno.cord
cc.cord <- as.numeric(gsub(".*: ","",pheno.cord$characteristics_ch1.11))
```


```{r}
# Create an empty data frame with gene names
pvals <- data.frame(Gene = rownames(edata.cord))

# Compute raw p-values using apply()
pvals$raw_pval <- apply(edata.cord, 1, function(gene_exprs){
  # Create a data frame where gene expression aligns with cotinine levels
  df.tmp <- data.frame(G= gene_exprs, CONC = cc.cord)
  
  # Fit linear model
  lm.tmp <- lm(CONC ~ G, df.tmp)
  
  # Extract p-value from ANOVA
  anova(lm.tmp)$"Pr(>F)"[1]
})

# Apply multiple testing corrections AFTER collecting p-values
pvals$bonferroni_pval <- p.adjust(pvals$raw_pval, method = "bonferroni") #if time - raw_pval / number samples
pvals$FDR_pval <- p.adjust(pvals$raw_pval, method = "fdr") #if time, try to replace with q-value script to find FDR

# create copy of edata + merge into copy
edata.cord.p <- as.data.frame(edata.cord)
edata.cord.p <- cbind(edata.cord, pvals[, -1])  # Drop Gene column, keep p-values

```

Order by raw, bonferroni, FDR + plot each

```{r}
# Order genes separately for each correction type
pvals_raw <- pvals[order(pvals$raw_pval), ]
pvals_bonferroni <- pvals[order(pvals$bonferroni_pval), ]
pvals_fdr <- pvals[order(pvals$FDR_pval), ]

# Scatter plots for each p-value type
#oldpar <- par(mfrow= c(1,3)) -- made plots too skinny
plot(-log10(pvals_raw$raw_pval), pch = 16, col = "blue", main = "Raw P-Values")
plot(-log10(pvals_bonferroni$bonferroni_pval), pch = 16, col = "red", main = "Bonferroni P-Values")
plot(-log10(pvals_fdr$FDR_pval), pch = 16, col = "green", main = "FDR P-Values")
#par(oldpar)

```
```{r}
#print top 10 pvals for each
pvals_raw[1:10,]
pvals_bonferroni[1:10,]
pvals_fdr[1:10,]
```

